DOC: adding recs for conda env file creation & plugin installation

lizgehret opened the pull request.

lizgehret commented:
cc @ebolyen - is this what you had in mind for conda install? and then just running `conda update` as needed with a newer env file (as the developer goes through their own 'releases' i.e. GH branches)?

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,163 @@ Some guidance on each of these approaches:
-Either of these approaches is totally fine.
-The former is an easier starting point.
+Either of these approaches is totally fine, with the following caveat.
+The former is an easier starting point, and will allow us to provide more troubleshooting assistance for any installation issues that users may encounter. While the latter provides you with more flexibility in your environment construction, our assistance with any install issues or environment conflicts that users may run into will be more limited (and will ultimately be your responsibility to tr ...
```
It's a bit of an odd convention, but we write this with one sentence per line. That just makes diffs easier to view, and the rendering engine ignores the single line breaks. Could you break this up so it's one sentence per line?

lizgehret replied:
Oh yeah, can do!

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,163 @@ Some guidance on each of these approaches:
+With the following terms defined:
+- 20XX.REL epoch corresponding to the QIIME 2 release version that your plugin is compatible with
+- <target-distribution> being the distribution your plugin should be utilized with (i.e. amplicon, metagenome)
+- <q2_my_plugin> being the name of your plugin
```
I've been calling this the `module_name` in the cookiecutter (see [here](https://github.com/caporaso-lab/cookiecutter-qiime2-plugin/blob/main/cookiecutter.json)) - unless you think there's a reason not to, can you use that term here too for consistency?

lizgehret replied:
yep, can do!

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,163 @@ Some guidance on each of these approaches:
+- <target-distribution> being the distribution your plugin should be utilized with (i.e. amplicon, metagenome)
+- <q2_my_plugin> being the name of your plugin
+- <your-github-org> either being the github organization your plugin is hosted under, or your personal github account name
+- <q2-my-plugin> being the name for your plugin on github
```
Similar to above, this is what I've been calling the plugin name.

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,163 @@ Some guidance on each of these approaches:
+
+2024.5
+- Environment file name:
+`2024.5-q2-myplugin-environment.yml`
```
```suggestion
`2024.5-q2-my-plugin-environment.yml`
```
(for consistency with the text above).

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
````diff
@@ -18,13 +18,163 @@ Some guidance on each of these approaches:
+Users' initial install command would look like this:
+
+```
+conda env create -n q2-myplugin -f https://raw.githubusercontent.com/myplugin-org/q2-myplugin/main/environment-files/2024.5-myplugin-environment.yml
````
Similar to my suggestion above, this isn't consistent with the environment names above. I recommend keeping it consistent, just so readers aren't wondering if there is some reason for the difference.

lizgehret commented:
Hey @gregcaporaso, just addressed your comments - I updated the standardized example to `<module_name>` and then used `<my_plugin>` and `<my-plugin>` in the specific examples. Let me know if that's what you had in mind, or if you think I should change them all to `<module_name>` and `<module-name>` - I have no preference, so whichever you think will be more clear for users!

gregcaporaso commented:
Thanks @lizgehret - is this one ready for final review, or are you still working on it?

lizgehret commented:
> Thanks @lizgehret - is this one ready for final review, or are you still working on it?

Still working on it @gregcaporaso - I'll add the second tutorial on creating a recurring GHA today, and let you know when it's ready for final review!

lizgehret commented:
Added a basic draft, still need to re-read what I wrote and add example GHAs.

lizgehret commented:
latest commit includes finalized GHA template for regular testing and description for weekly builds. just need to add that workflow in distributions and then write out a template and this will be g2g!

lizgehret commented:
Okay @gregcaporaso this should be ready for review whenever you have the chance!

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,259 @@ Some guidance on each of these approaches:
+2. Installing your plugin using the Tiny Distribution and any custom required plugins (not recommended)
+
+If you are working on a unique plugin that is not compatible with one of our existing distributions (amplicon, metagenome) that has a few specific q2 plugin dependencies, you'll utilize a similar approach to install option 1 - just with a more customized environment file.
+As a reminder, while this approach is fairly straightforward to implement, we don't recommend this (if at all possible) because this will be more difficult for us to assist with and help users to troubleshoot.
```
```suggestion
As a reminder, while this approach is fairly straightforward to implement, we don't recommend this if the option presented above is possible for your plugin, because this will be more difficult for us to assist with and help users to troubleshoot.
```

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,259 @@ Some guidance on each of these approaches:
+
+1. Continuous Integration (CI) to regularly run your plugin's unit tests within the type of QIIME 2 environment where it should be used (i.e. Amplicon, Metagenome, Tiny plus custom selections).
+
+To implement this, you'll need to create a Github Action (GHA) that will be triggered each time you make a commit to your repository - either through a pull request (PR) or a direct commit to one of your remote branches.
```
Should this be templated in, in place of the one we already have in the cookiecutter?

lizgehret replied:
I've been going back and forth about this, and I'm kind of leaning towards having them implement this manually vs. having it get automatically added when they create their plugin. I think it's more beneficial to have them set this up when their plugin is actually ready to be automatically tested vs. before they have any functionality laid out in their repository. I could totally be convinced otherwise, but I think it might help them understand this process a bit better if they have to set it up themselves (esp since they need to create a matching environment file that probably won't exist right away). I also worry that if they do have this setup too soon and the action is failing, it'll be more confusing for them.

gregcaporaso commented on book/plugins/how-to-guides/maximize-compatibility.md:
```diff
@@ -18,13 +18,259 @@ Some guidance on each of these approaches:
+
+You're welcome to set up any additional actions that you feel will be beneficial to your plugin and general development workflow - but the actions outlined above are what we recommend having in order to maintain an active and healthy plugin within the QIIME 2 ecosystem.
+
 You can request feedback on your plugin as a whole from more experienced QIIME 2 developers by reaching out through the {{ developer_discussion }}.
```
This should be a new section heading

lizgehret commented on book/plugins/how-to-guides/maximize-compatibility.md:
````diff
@@ -18,13 +18,259 @@ Some guidance on each of these approaches:
+environment-files/
+```
+
+Within this directory, you'll create environment file(s) for current and/or past installable versions of your plugin. You can name them something like:
+
+```
+20XX.REL-<module-name>-environment.yml
+```
````
I like it!

lizgehret commented:
> @lizgehret, I'm wondering if maybe this makes more sense to go in the `Distribute plugins on GitHub` How-to - thoughts on that?

Ah that makes sense, yeah we can move all of this over there if that feels like a better fit!

I'll address all of the formatting stuff today; let me know if you want me to make any additional adjustments after that or if you'd just prefer to make any other changes you want to see.

lizgehret commented:
Okay @gregcaporaso formatting updates are all done - I'll leave it as-is for you to move around (if you think it should go under the `Distribute on GH` section) or make any other changes you think make sense!

gregcaporaso commented:
Closing in favor of the interactive rebased version of this PR. #119
